### PR TITLE
Bugfix: allow missing modules if they are blacklisted

### LINKED ISSUE #13346: spack env loads fails if there are blacklisted modules
I have an environment with blacklisted modules. Running `spack env loads` fails if run with blacklisted modules.

```console
$ spack env loads --module-type tcl --dependencies --input-only
==> Error: No module available for package py-numpy@1.17.2%gcc@9.1.0+blas+lapack arch=linux-centos7-sandybridge ^bzip2@1.0.6%gcc@9.1.0+shared arch=linux-centos7-sandybridge ^diffutils@3.7%gcc@9.1.0 arch=linux-centos7-sandybridge ^expat@2.2.9%gcc@9.1.0+libbsd arch=linux-centos7-sandybridge ^gdbm@1.18.1%gcc@9.1.0 arch=linux-centos7-sandybridge ^gettext@0.20.1%gcc@9.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-sandybridge ^intel-mkl@2019.3.199%gcc@9.1.0~ilp64+shared threads=none arch=linux-centos7-sandybridge ^libbsd@0.9.1%gcc@9.1.0 arch=linux-centos7-sandybridge ^libffi@3.2.1%gcc@9.1.0 arch=linux-centos7-sandybridge ^libiconv@1.16%gcc@9.1.0 arch=linux-centos7-sandybridge ^libxml2@2.9.9%gcc@9.1.0~python arch=linux-centos7-sandybridge ^ncurses@6.1%gcc@9.1.0~symlinks~termlib arch=linux-centos7-sandybridge ^openssl@1.0.2k%gcc@9.1.0+systemcerts arch=linux-centos7-sandybridge ^pkgconf@1.6.3%gcc@9.1.0 arch=linux-centos7-sandybridge ^py-setuptools@41.4.0%gcc@9.1.0 arch=linux-centos7-sandybridge ^python@3.7.3%gcc@9.1.0+bz2+ctypes+dbm+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4~uuid+zlib arch=linux-centos7-sandybridge ^readline@8.0%gcc@9.1.0 arch=linux-centos7-sandybridge ^sqlite@3.30.0%gcc@9.1.0~column_metadata+fts~functions~rtree arch=linux-centos7-sandybridge ^tar@1.32%gcc@9.1.0 arch=linux-centos7-sandybridge ^xz@5.2.4%gcc@9.1.0 arch=linux-centos7-sandybridge ^zlib@1.2.11%gcc@9.1.0+optimize+pic+shared arch=linux-centos7-sandybridge at /opt/packages/gpjohnsn/opt/ssoft/modules/packages/2019.1/linux-centos7-sandybridge/py-numpy/1.17.2_gcc-9.1.0-python-3.7.3
```
This used to work by putting the packages that did have modules in the loads file.

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes #13346.

For `spack module loads` and `spack module find`, allow missing upstream modules (print a debug message vs. raising an error); also skip local modules if they are blacklisted.

Old behavior:

* `spack module find` will report an error if the root spec's module or any dependency module does not exist (regardless of whether it is blacklisted)
* `spack module loads` works the same

New behavior: the short version is that the module finding is now more lenient (especially for blacklisted modules):

* `spack module find` will report an error if the root spec's module does not exist, unless
  * the module for the spec is blacklisted
    * For the root package, a debug message is output so that the user could check this with `spack -d...`
    * For any dependency package (when requesting dependency modules with `spack module find`) a message will not be reported (it's an error if the module is not blacklisted, but if it is blacklisted, the module is silently skipped)
  * the spec is associated with an upstream
    * instead of throwing exceptions when the module isn't find, the logic now prints out a debug-level message so the user has a hint of what is missing if they were expecting to see a module file for an upstream package
* `spack module loads` works exactly the same way as the new `spack module find`, but informational messages are not generated for blacklisted packages

### CODE REVIEW

**Comment by scheibelp** [comment]
@tgamblin this should add tests (e.g. for `spack module find` when there is a blacklist) but do you agree with the intended behavior in the PR description?

**Comment by tgamblin** [comment]
@glennpj: does this fix #13346 for you?

**Comment by glennpj** [comment]
> @glennpj: does this fix #13346 for you?

Yes, it does.

File: lib/spack/spack/modules/common.py [hunk 1]
```diff
@@ -312,31 +312,51 @@ def upstream_module(self, spec, module_type):
...
-def get_module(module_type, spec, get_full_path):
+def get_module(module_type, spec, get_full_path, required=True):
+    """Retrieve the module file for a given spec and module type. If
+    ``required`` is set, then an exception will be raised if the module is
+    not found and the Spec is not blacklisted in the module configuration.
+    The upstream module index does not record whether a module is blacklisted:
+    if it is missing it is assumed to have been omitted purposefully. If the
+    module is not available (and it is not required or it is blacklisted), then
+    this function returns ``None``.
+    """
```
Comment (alalazo) [inline lib/spack/spack/modules/common.py]: Minor OCD. Can we reformat to respect the [Google Style Guide](http://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings), as we do in many other places?

Comment (tgamblin) [reply lib/spack/spack/modules/common.py]: Is there a checker for this?

Comment (alalazo) [reply lib/spack/spack/modules/common.py]: Not sure, really. There's a checker for [PEP257](http://www.pydocstyle.org/en/4.0.1/release_notes.html), but they seems to have only "initial support" for Google's conventions.

**Comment by scheibelp** [comment]
@tgamblin changes are now addressed:

* explicit return of `None`
* updated docstring (per https://github.com/spack/spack/pull/13540#discussion_r353805350)
* `loads` output includes comments for missing (blacklisted/upstream) modules (note that the comment mentions that it could be blacklisted or that it could be missing from an upstream but doesn't distinguish) as well as an overall warning message

**Pull request merged**